### PR TITLE
fixing agent pool upgrade logic

### PR DESCRIPTION
**Pull request by dmitsh**
fixing agent pool upgrade logic
 - checking provisioning state of existing VMs
 - do not create extra VM if all nodes have been already upgraded
 - adding more logs

### CODE REVIEW

**Comment by weinong** [comment]
can we add tests?

File: pkg/operations/kubernetesupgrade/upgrader.go [hunk 1]
```diff
@@ -221,98 +233,126 @@ func (ku *Upgrader) upgradeAgentPools() error {
...
 				vmProvisioningState = *vm.VirtualMachineProperties.ProvisioningState
 			}
+			agentIndex, _ := armhelpers.GetVMNameIndex(vm.StorageProfile.OsDisk.OsType, *vm.Name)
 
-			if vmProvisioningState != string(api.Succeeded) {
-				ku.logger.Infof("Deleting agent VM %s in provisioning state %s\n", *vm.Name, vmProvisioningState)
+			switch vmProvisioningState {
+			case "Creating":
+				fallthrough
+			case "Updating":
```
Comment (JackQuincy) [inline pkg/operations/kubernetesupgrade/upgrader.go]: these can be written more compactly:
 case "Creating", "Updating", "Succeeded": 
Really a matter of choice though. @jackfrancis @amanohar any opinion on which style we want?

Comment (jackfrancis) [reply pkg/operations/kubernetesupgrade/upgrader.go]: +1 for the succinct style. See the "Multiple Cases" here:

https://github.com/golang/go/wiki/Switch

Comment (dmitsh) [reply pkg/operations/kubernetesupgrade/upgrader.go]: sure, I will change that.

**Comment by dmitsh** [comment]
@weinong We would need to overhaul current ARM simulation framework. I will create a task and we will do that in a separate PR.

File: pkg/operations/kubernetesupgrade/upgradeagentnode.go [hunk 2]
```diff
@@ -112,30 +113,27 @@ func (kan *UpgradeAgentNode) Validate(vmName *string) error {
...
-	go func() {
-		for {
+	retryTimer := time.NewTimer(time.Millisecond)
+	timeoutTimer := time.NewTimer(timeout)
+	for {
+		select {
+		case <-timeoutTimer.C:
+			retryTimer.Stop()
+			kan.logger.Errorf("Node was not ready within %v", timeout)
+			return fmt.Errorf("Node was not ready within %v", timeout)
```
Comment (JackQuincy) [inline pkg/operations/kubernetesupgrade/upgradeagentnode.go]: [nit] err := fmt.Errorf("Node was not ready within %v", timeout")
kan.logger.Error(err.Error())
return err

**Pull request merged**